kernel/fs.c

```diff
@@ -336,14 +340,57 @@ int fs_file_write(struct fs_file *file, const char *buffer, uint32_t length, uin
 	return total;
 }
 
-int fs_file_get_dimensions(struct fs_file *f, int * dims, int n) 
+int fs_dirent_dup(struct fs_dirent *src, struct fs_dirent *dst) {
+	char *buffer = kmalloc(4096);
+	memset(buffer,0,4096);
+	printf("Reading directory... ");
+	int length = fs_dirent_readdir(src, buffer, 4096);
+	printf("Done.\n");
+	if (length <= 0) {
+		return length;
+	}
+	char *name = buffer;
+	while (name && (name - buffer) < length) {
+		if (strcmp(name,".") == 0 || (strcmp(name, "..") == 0)) {
+			name += strlen(name) + 1;
+			continue;
+		}
+		printf("Copying %s...\n", name);
+		struct fs_dirent *new_src, *new_dst;
+		new_src = fs_dirent_lookup(src, name);
+		char temp[1];
+		if (fs_dirent_readdir(new_src, temp, 1) == KERROR_NOT_A_DIRECTORY) { //file
```

Comment:
This logic isn't quite right: you will take the else branch if the readdir succeeds *or* fails for some reason other than `NOT_A_DIRECTORY`.  I think you want this:

```
r = readdir(...);
if(r is success) {
   // handle the directory 
} else if(r is failure indicating a file) {
   // handle a file
} else {
   // handle other error
}
```